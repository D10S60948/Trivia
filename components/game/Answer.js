import React, { Component } from 'react';
import { StyleSheet, TouchableHighlight, Text, ImageBackground } from 'react-native';
import { connect } from 'react-redux';
import { addPoints, resetScore, reduceLife, editMessage, setVisibility } from '../../redux/actions';
import { correctAnswerAlerts, wrongAnswerAlerts } from '../../assistance/alerts';
import { makeSound } from '../../assistance/makeSounds';

export class Answer extends Component {

    handleCorrectAnswer(index) {
        if(this.props.soundsOn) makeSound(this.props.sounds.correct)
        this.props.editMessage("CORRECT !\n\n" + correctAnswerAlerts[index])
        this.props.questionType == 'selection' ? this.props.addPoints(10) : this.props.addPoints(5)
    }

    handleWrongAnswer(notificationContentIndex) {
        if(this.props.livesLeft == 1) {
            this.props.editMessage('GAME OVER')
        }
        else {
            if(this.props.soundsOn) makeSound(this.props.sounds.wrong)
            let message = this.props.livesLeft == 2 ? 'Wrong again..\n\nLast chance..' : "Wrong answer..\n\n" + wrongAnswerAlerts[notificationContentIndex]
            this.props.editMessage(message)
        }
        this.props.reduceLife()
    }

    getAnswerByIdentifier() {
        switch(this.props.questionIdentifier) {
            case 'flag':
                return this.props.correctAnswer.name
            case 'capital':
                return this.props.correctAnswer.capital
            case 'population':
                return this.props.correctAnswer.population.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",")
            case 'region':
                return this.props.correctAnswer.region
        }
    }

    handleAnswer = () => {
        let toBeComparedWith = this.props.questionType == 'selection' ? this.getAnswerByIdentifier() : this.props.correctAnswer.trueOrFalse
        let notificationContentIndex = Math.floor(Math.random() * 4)
        
        if(this.props.content == toBeComparedWith)  {
            this.handleCorrectAnswer(notificationContentIndex) 
        }
        else {
            this.handleWrongAnswer(notificationContentIndex)
        }
        this.props.setVisibility(true)   
    }
    
    render() {
        return (
            <ImageBackground 
                source={require('../../images/backgrounds/questionBG.png')} 
                style={styles.background} 
                imageStyle={{opacity: 0.5}}
                resizeMode='stretch'
                >
                <TouchableHighlight 
                style={this.props.questionType == 'selection' ? styles.container : {...styles.container, alignItems: 'center'}} 
                underlayColor='lightslategrey'
                onPress={this.handleAnswer}
                >
                    <Text style={styles.text}>
                        {this.props.content}
                    </Text>
                </TouchableHighlight>
            </ImageBackground>

        );
    }
}

const styles = StyleSheet.create({
  container: {
    width: '100%',
    height: '100%',
    justifyContent: 'center',
    borderRadius: 15,
  },
  text: {
    paddingHorizontal: 40,
    fontFamily: 'Signika',
    fontSize: 17
  },
  background: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    marginVertical: 3
  }
});

function mapStateToProps(state) {
    return {
        selectedAnswers: state.play.selectedAnswers,
        correctAnswer: state.play.correctAnswer,
        isReady: state.play.ready,
        questionType: state.play.question.type,
        questionIdentifier: state.play.question.identifier,
        livesLeft: state.lives,
        sounds: state.notifications.sounds,
        soundsOn: state.notifications.sounds.isOn
    };
}

export default connect(mapStateToProps, { addPoints, resetScore, reduceLife, editMessage, setVisibility })(Answer)
