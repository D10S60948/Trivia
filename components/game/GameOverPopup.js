import React, { Component } from 'react'
import { View, TouchableHighlight, Text, Modal, StyleSheet, AsyncStorage, TextInput } from 'react-native'
import {  setNotReady, resetAnwersAndQuestion, setVisibility, resetLives, resetScore } from '../../redux/actions'
import { connect } from 'react-redux'
import { withNavigation } from 'react-navigation'

class GameOverPopup extends Component {

    constructor(props) {
        super(props)
        this.state = {
            highScore: false,
            userName: '',
            text: ''
        }
        this.addScoreToHighestScores = this.addScoreToHighestScores.bind(this)
        this.resetDataAndSaveScore = this.resetDataAndSaveScore.bind(this)
        this.toMainPage = this.toMainPage.bind(this)
        this.restartGame = this.restartGame.bind(this)
        this.getUserName = this.getUserName.bind(this)
    }

    async addScoreToHighestScores(highestScores, currenctScore) {
        let listLength = highestScores.length
        let wasAdded = false

        for(let i=0; i<listLength; i++) {
            if(currenctScore.score >= highestScores[i].score) {
                highestScores.splice(i, 0, currenctScore)
                if(listLength == 10) {
                    highestScores.pop()
                }
                wasAdded = true
                break
            }
        }
        if(!wasAdded) {
            highestScores.push(currenctScore)
        }
        await AsyncStorage.setItem('highScores', JSON.stringify(highestScores))
    }

    async resetDataAndSaveScore() {
        if(this.state.highScore) {
            await this.addScoreToHighestScores(this.props.highestScores, {score: this.props.score, name: this.state.text})
        }
        this.props.resetScore()
        this.props.setVisibility(false)
        this.props.resetLives()
    }

    async toMainPage() {
        await this.resetDataAndSaveScore()
        this.props.navigation.navigate('MainPage')
    }

    async restartGame() {
        await this.resetDataAndSaveScore()
        this.props.navigation.navigate('GamePage')
    }

    componentWillMount() {
        let highestScoresList = this.props.highestScores
        if(highestScoresList.length < 10 || this.props.score >= highestScoresList[highestScoresList.length-1].score) {
            this.setState({ highScore: true })
        }
    }

    getUserName() {
        return (
            <View style={{ justifyContent: 'center', alignItems: 'center', flex: 1 }}> 
                <Text style={styles.text}>Your Score was high!</Text>
                <Text style={styles.text}>Please insert name or nickname</Text>
                <TextInput
                    onChangeText={(text) => this.setState({text})}
                    value={this.state.text}
                    style={{...styles.text, color: 'black', padding: 7, backgroundColor: 'lightgray', width: '80%'}}
                />
            </View>
        )
    }

    render() {
        return (
            <Modal
            animationType="fade"
            onRequestClose={() => { console.log('Modal has been closed.') }}
            transparent={true}
            visible={this.props.visibility}
            >
                <View style={styles.container}>
                    <View style={styles.box}>
                        <View style={{ justifyContent: 'center', alignItems: 'center', flex: 1 }}> 
                            <Text style={styles.text}>GAME OVER</Text>
                        </View>
                        { this.state.highScore ? this.getUserName() : null }
                        <View style={{ justifyContent: 'center', alignItems: 'center', flex: 1, flexDirection: 'row' }}> 
                            <TouchableHighlight onPress={() => this.toMainPage()} style={styles.restartButton}>
                                <Text style={{ fontFamily: 'Signika'}}>Back to menu</Text>
                            </TouchableHighlight>
                            <TouchableHighlight onPress={() => this.restartGame()} style={styles.restartButton}>
                                <Text style={{ fontFamily: 'Signika'}}>Try again</Text>
                            </TouchableHighlight>
                        </View>
                    </View>
                </View>
            </Modal>
        )
    }
}

const styles = StyleSheet.create({
    container: {
        flex:1,
        backgroundColor: 'rgba(0,0,0,0.5)',
        justifyContent: 'center',
        alignItems: 'center'
    },
    box: {
        backgroundColor: 'black',
        height: '92%',
        width: '90%',
        justifyContent: 'space-around',
        alignItems: 'center',
        borderRadius: 20
    },
    restartButton: {
        backgroundColor: 'grey',
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: 5,
        paddingVertical: 5,
        paddingHorizontal: 25
    },
    text: {
        fontFamily: 'Signika', 
        color: 'white', 
        marginVertical: 15,
        textAlign: 'center'
    }
})

function mapStateToProps(state) {
    return {
        visibility: state.notifications.visibility,
        message: state.notifications.message,
        livesLeft: state.lives,
        score: state.score.currentScore,
        highestScores: state.score.highestScoreList
    };
}

export default connect(mapStateToProps, { setNotReady, resetAnwersAndQuestion, setVisibility, resetLives, resetScore })(withNavigation(GameOverPopup))
