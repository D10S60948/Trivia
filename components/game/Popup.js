import React, { Component } from 'react'
import { View, TouchableHighlight, Text, Modal, StyleSheet, AsyncStorage } from 'react-native'
import {  setNotReady, resetAnwersAndQuestion, setVisibility, resetLives, resetScore } from '../../redux/actions'
import { connect } from 'react-redux'
import { withNavigation } from 'react-navigation'

export class Popup extends Component {

    handleClosePopup() {
        this.props.setNotReady()
        this.props.resetAnwersAndQuestion()
        this.props.setVisibility(false)
    }

    componentWillMount() {
        if(this.props.livesLeft == 0) {
            this.setState({ isGameOver: true })
        }
    }

    render() {
        return (
            <Modal
            animationType="fade"
            onRequestClose={() => { console.log('Modal has been closed.') }}
            transparent={true}
            visible={this.props.visibility}
            >
                <View style={styles.container}>
                    <View style={styles.box}>
                        <Text style={styles.text}>{this.props.message}</Text>
                        <TouchableHighlight onPress={() => this.handleClosePopup()} style={styles.closeModal}>
                            <Text style={{ fontFamily: 'Signika'}}>close</Text>
                        </TouchableHighlight>
                    </View>
                </View>
            </Modal>
        )
    }
}

const styles = StyleSheet.create({
    container: {
        flex:1,
        backgroundColor: 'rgba(0,0,0,0.5)',
        justifyContent: 'center',
        alignItems: 'center'
    },
    box: {
        backgroundColor: 'black',
        height: '32%',
        width: '80%',
        justifyContent: 'space-around',
        alignItems: 'center',
        borderRadius: 20
    },
    closeModal: {
        backgroundColor: 'grey',
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: 10,
        paddingVertical: 7,
        paddingHorizontal: 10
    },
    text: {
        fontFamily: 'Signika', 
        color: 'white', 
        marginVertical: 25,
        textAlign: 'center'
    }
})

function mapStateToProps(state) {
    return {
        visibility: state.notifications.visibility,
        message: state.notifications.message,
        livesLeft: state.lives,
        score: state.score.currentScore,
        highestScores: state.score.highestScoreList
    };
}

export default connect(mapStateToProps, { setNotReady, resetAnwersAndQuestion, setVisibility, resetLives, resetScore })(withNavigation(Popup))
