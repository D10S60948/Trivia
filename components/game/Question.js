import React, { Component } from 'react'
import { StyleSheet, ImageBackground } from 'react-native'
import { connect } from 'react-redux'
import { resetAnwersAndQuestion } from '../../redux/actions'
import FlagQuestion from './questions/FlagQuestion'
import CapitalQuestion from './questions/CapitalQuestion'
import LanguagesQuestion from './questions/LanguagesQuestion'
import BordersQuestion from './questions/BordersQuestion'
import PopulationQuestion from './questions/PopulationQuestion'
import RegionQuestion from './questions/RegionQuestion'

export class Question extends Component {

    setQuestion() {
        let identifierType = this.props.question.identifier
        let correctAnswerCountry = this.props.correctAnswer

        switch(identifierType) {
            case 'flag':
                return <FlagQuestion flag={ correctAnswerCountry.emoji } />
            case 'capital':
                return <CapitalQuestion countryName={correctAnswerCountry.name} />
            case 'language':
                if(correctAnswerCountry.languages.length == 0) this.props.resetAnwersAndQuestion()
                return <LanguagesQuestion languages={correctAnswerCountry.languages} name={correctAnswerCountry.name} />
            case 'borders':
                return <BordersQuestion borders={correctAnswerCountry.borders} name={correctAnswerCountry.name} countries={this.props.allCountries} />
            case 'population':
                return <PopulationQuestion name={correctAnswerCountry.name} />
            case 'region':
                return <RegionQuestion name={correctAnswerCountry.name} region={correctAnswerCountry.region} />
        }
    }
    
    render() {
        return (
            <ImageBackground 
            source={require('../../images/backgrounds/questionBG.png')} 
            style={styles.container} 
            imageStyle={{opacity: 0.5}}
            resizeMode='stretch'
            >
                { this.props.ready && this.setQuestion() }
            </ImageBackground>
        );
    }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center', 
    flexDirection: 'column', 
    flexWrap: 'wrap'
  }
});

function mapStateToProps(state) {
    return {
        ready: state.play.ready,
        question: state.play.question,
        correctAnswer: state.play.correctAnswer,
        allCountries: state.play.countries
    };
}

export default connect(mapStateToProps, { resetAnwersAndQuestion })(Question)