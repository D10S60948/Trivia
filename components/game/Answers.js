import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import Answer from './Answer';
import { connect } from 'react-redux'

class Answers extends Component {

    getAnswerType(answer, key) {
        switch(this.props.questionType.identifier) {
            case 'flag':
                return <Answer key={key} content={answer.name} />
            case 'capital':
                return <Answer key={key} content={answer.capital} />
            case 'population':
                return <Answer key={key} content={answer.population} />
            case 'region':
                return <Answer key={key} content={answer.region} />
        }
    }

    getAnswers() {
        if(this.props.questionType.type == 'selection') {
            return (
                <View style={styles.container}>
                    {this.props.selectedAnswers.map((answer, key) => this.getAnswerType(answer, key))}
                </View>
            )
        }
        else {
            return (
                <View style={styles.container}>
                    <Answer content={'True'} />
                    <Answer content={'False'} />
                </View>
            )
        }
    }

    render() {
        return (
            <View style={styles.background} >
                { this.props.isReady && this.getAnswers() }
            </View>
        );
    }
}

const styles = StyleSheet.create({
    background: {
    flex: 2,
    justifyContent: 'center',
    alignItems: 'center'
  },
  container: {
    height: '80%',
    width:'98%',
    justifyContent: 'space-around',
  }
});

function mapStateToProps(state) {
    return {
        selectedAnswers: state.play.selectedAnswers,
        isReady: state.play.ready,
        questionType: state.play.question
    };
}

export default connect(mapStateToProps, null)(Answers)
