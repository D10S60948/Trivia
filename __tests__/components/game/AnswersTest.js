import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { Answers } from '../../../components/game/Answers';

test('"Answers" component renders correctly', () => {
    const snapshot = renderer.create(<Answers/>).toJSON();
    expect(snapshot).toMatchSnapshot();
});
