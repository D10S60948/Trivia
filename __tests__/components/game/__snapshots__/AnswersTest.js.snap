// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"Answers" component renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 2,
      "justifyContent": "center",
    }
  }
/>
`;
